File: android-core/src/main/java/com/mparticle/internal/messages/MPEventMessage.java
package com.mparticle.internal.messages;

import android.location.Location;

import androidx.annotation.Nullable;

import com.mparticle.MParticle;
import com.mparticle.internal.Constants;
import com.mparticle.internal.InternalSession;

import org.json.JSONException;

public class MPEventMessage extends BaseMPMessage {

    protected MPEventMessage(Builder builder, InternalSession session, @Nullable Location location, long mpId) throws JSONException {
        super(builder, session, location, mpId);
    }

    public static class Builder extends BaseMPMessageBuilder {

        public Builder(String messageType) {
            super(messageType);
        }

        public BaseMPMessageBuilder customEventType(MParticle.EventType eventType) {
            try {
                put(Constants.MessageKey.EVENT_TYPE, eventType);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            return this;
        }

        @Override
        public BaseMPMessage build(InternalSession session, Location location, long mpId) throws JSONException {
            return new MPEventMessage(this, session, location, mpId);
        }
    }
}

File: android-core/src/main/kotlin/com/mparticle/internal/messages/MPEventMessage.kt
package com.mparticle.internal.messages

import android.location.Location
import com.mparticle.MParticle
import com.mparticle.internal.Constants
import com.mparticle.internal.InternalSession
import org.json.JSONException

class MPEventMessage protected constructor(builder: Builder, session: InternalSession, location: Location?, mpId: Long) :
    BaseMPMessage(builder, session, location, mpId) {
    class Builder(messageType: String) : BaseMPMessageBuilder(messageType) {
        fun customEventType(eventType: MParticle.EventType): BaseMPMessageBuilder {
            try {
                put(Constants.MessageKey.EVENT_TYPE, eventType)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return this
        }

        @Throws(JSONException::class)
        override fun build(session: InternalSession, location: Location?, mpId: Long): BaseMPMessage {
            return MPEventMessage(this, session, location, mpId)
        }
    }
}